### ISSUE TITLE
Reduce the usage of GroupLookup for window

### ISSUE THREAD

**Issue by jsternberg** [influxdata/flux#1036]
This issue is created because of the work in #1020 and is related to #1035. Window generates too many group keys which seems to overload `GroupLookup`. While it is important to optimize `GroupLookup` for the general health of the system, window may be able to use a different method to reduce its problems with `GroupLookup`. Instead of making a new group key for each window and storing it in group lookup, a key can be used for the group key's parent and all of the windows within that can be stored in a more simple format that contains the different intervals. It should then be possible to implement a custom `DataCache` that iterates over the `GroupLookup` and then iterates over each interval instead of storing them individually inside of the `GroupLookup`.

**Comment by jsternberg** [comment]
I think I know specifically why `GroupLookup` works so poorly in window compared to the other methods. It seems that we've got a worst case scenario of using `GroupLookup` that happens every time we run `window()`. This worst case scenario is documented in #1035 and I'm planning to address it there.

The reason why this seems to happen is that `GroupLookup` is designed to keep things lazily sort, but we invoke the methods in a way where it never lazily sorts anything. If you called `Set` over and over again (which is functionally what happens since the group key is unique when calling process and the generated key will thus likely never conflict with a future table unless specifically designed to conflict), then the sorting will only happen when `Range` or `Lookup` is called. But, the `TableBuilderCache` always invokes `Lookup` followed by `Set` when inserting new tables which triggers a sort, if necessary, on every time we add a new table.

This isn't a problem for a lot of transformations. But, this is a problem for window because of how it modifies the group key. If you were to take the following with `window(every: 1m)`:

```
_measurement=cpu,_field=usage_user,_start=2019-01-01T00:00:00Z,_stop=2019-01-01T00:02:00Z,t0=a
_measurement=cpu,_field=usage_user,_start=2019-01-01T00:00:00Z,_stop=2019-01-01T00:02:00Z,t0=b
```

This is pretty standard. It has one tag. The window transformation receives each of these tables in order and produces this for the first table.

```
_measurement=cpu,_field=usage_user,_start=2019-01-01T00:00:00Z,_stop=2019-01-01T00:01:00Z,t0=a
_measurement=cpu,_field=usage_user,_start=2019-01-01T00:01:00Z,_stop=2019-01-01T00:02:00Z,t0=a
```

Then it processes the second table. Notice how when it inserts the new tables for this second series, we are _not_ appending. It requires a full sort on every insert.

```
_measurement=cpu,_field=usage_user,_start=2019-01-01T00:00:00Z,_stop=2019-01-01T00:01:00Z,t0=a
_measurement=cpu,_field=usage_user,_start=2019-01-01T00:00:00Z,_stop=2019-01-01T00:01:00Z,t0=b
_measurement=cpu,_field=usage_user,_start=2019-01-01T00:01:00Z,_stop=2019-01-01T00:02:00Z,t0=a
_measurement=cpu,_field=usage_user,_start=2019-01-01T00:01:00Z,_stop=2019-01-01T00:02:00Z,t0=b
```

In this way, we get a worst case scenario. My proposal is simple. It is to change the internal structure of the group lookup so that it acts similar to a trie and also to modify the `Range` method to make the structure more append friendly. This will change the order of the output so if that is important not to change, I want to hear that it is important not to change.k

```
// Old:
// Range will iterate over all groups keys in sorted order.
// Range must not be called within another call to Range.
// It is safe to call Set/Delete while ranging.
// New:
// Range will iterate over all groups keys in the order they were added.
// Range must not be called within another call to Range.
// It is safe to call Set/Delete while ranging.
```

**Comment by nathanielc** [comment]
I can add some history to why it is the way it is. That doesn't mean we can't change it, but the context is important.

1. The sort order was initially designed to help the join transformation, I don't think this is the case any longer and it may be safe to change.
2. We wanted the output of tables to be in a stable order. Initially group lookup was a Go map and each time you queried data it would be returned in a different order. This even effects the csv output as it doesn't do any sorting itself and so the csv tables would change order making it hard to debug.
3. The current sort order is chosen because it follows the sort order that storage uses. With the optimization to only sort if the new element is not the largest means that when reading tables from storage no sorts need to occur. 

I am most concerned about the second point, where we want Flux to be stable such that it is easy to debug and reason about returned results. The particular sort order or that it is sorted at all is not as important to me. But having a stable order is. 


TL;DR at first pass it seems like it would be fine to change Range to not sort but rather preserve the insertion order. The question is whether insertion order is stable.

**Comment by nathanielc** [comment]
And one last thing. I think the SPEC says that the tables will be sorted on the group key. If we make this change we need to update that part of the SPEC.

In general I would like to not commit to a specific sort order in the SPEC as that has significant performance implications.

**Comment by jsternberg** [comment]
The second is the big one for me. I think the emphasis should be on _stable_ though which makes using a map as the sole backing to be incorrect. The proposal would be the order they were added. This means for most transformations we would not change the order with which they were emitted and others would always be deterministic.

If you're fine with "deterministic, but not sorted" I can change that in the spec and then begin working on this change.

**Comment by nathanielc** [comment]
Yeah "deterministic but not sorted" is what I mean by stable. Yes, lets start work on that. The part I don't know is if the transport of tables between transformations is stable such that we can rely on it without some kind of sorting behavior.

**Comment by jsternberg** [comment]
I'll be addressing this in #1035 and putting my notes about benchmarking there since it is a more general problem.

**Comment by nathanielc** [comment]
@jsternberg  Has this issue been addressed completely?

**Comment by jsternberg** [comment]
I think so. The other issues with window should be addressed separately.
